src/lib.rs

```diff
@@ -497,6 +527,24 @@ impl LightningNode {
         amount_sat: u64,
         _metadata: String,
     ) -> PayResult<()> {
+        match self.rt.handle().block_on(parse(&invoice)) {
+            Ok(InputType::Bolt11 { invoice }) => {
+                let user_preferences = self.user_preferences.lock().unwrap().clone();
+                let exchange_rates = self.task_manager.lock().unwrap().get_exchange_rates();
+                self.data_store
+                    .lock()
+                    .unwrap()
+                    .store_payment_info(
+                        &invoice.payment_hash,
+                        user_preferences,
+                        exchange_rates,
+                        None,
+                    )
+                    .map_to_permanent_failure("Failed to persist payment info")
+            }
+            _ => Err(invalid_input("Invalid invoice")),
+        }?;
+
```

Comment:
`pay_invoice()` and `pay_open_invoice()` now duplicate a lot of code. We should probably create a private function that contains that code and is being called by `pay_invoice()` and `pay_open_invoice()`. Maybe in another PR.

Reply:
There's still some duplicated code, but it got improved in [c90d94e](https://github.com/getlipa/lipa-lightning-lib/pull/628/commits/c90d94ef4ad97535821a2a4f9fb5af7259137126). I would leave more improvements, if any, for later PRs.